clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -775,20 +789,46 @@ scheduler_tmpl_t<RecordType, ReaderType>::advance_region_of_interest(int output_
     return sched_type_t::STATUS_SKIPPED;
 }
 
+template <typename RecordType, typename ReaderType>
+void
+scheduler_tmpl_t<RecordType, ReaderType>::set_cur_input(int output_ordinal,
+                                                        int input_ordinal)
+{
+    assert(output_ordinal >= 0 && output_ordinal < static_cast<int>(outputs_.size()));
+    assert(input_ordinal < static_cast<int>(inputs_.size()));
+    if (outputs_[output_ordinal].cur_input >= 0)
+        ready_.push(outputs_[output_ordinal].cur_input);
+    outputs_[output_ordinal].cur_input = input_ordinal;
+    if (input_ordinal < 0)
+        return;
+    inputs_[input_ordinal].instrs_in_quantum = 0;
+}
+
 template <typename RecordType, typename ReaderType>
 typename scheduler_tmpl_t<RecordType, ReaderType>::stream_status_t
 scheduler_tmpl_t<RecordType, ReaderType>::pick_next_input(int output_ordinal)
 {
-    if (options_.mapping == MAP_TO_ANY_OUTPUT) {
-        // TODO i#5843: Implement synthetic scheduling with instr/time quanta.
-        // These will require locks and central coordination.
-        // For now we only support a single output.
-        return sched_type_t::STATUS_EOF;
-    }
-    int index = outputs_[output_ordinal].cur_input;
+    if (options_.mapping == MAP_TO_ANY_OUTPUT)
+        std::lock_guard<std::mutex> guard(sched_lock_);
```

Comment:
The scope of this lock_guard contains only this declaration. Is it useful?

Reply:
No this is not what we want.  Replaced with `std::unique_lock` for a conditional lock that still auto-unlocks on any return path:
```C
    bool need_lock = (options_.mapping == MAP_TO_ANY_OUTPUT);
    auto scoped_lock = need_lock ? std::unique_lock<std::mutex>()
                                 : std::unique_lock<std::mutex>(sched_lock_);
```